feat: improve AudioPlayer

Halkiion opened the pull request:
Fixed several bugs such as:

- Wrong audio playing
- Crashing when switching guild or channel whilst audio is playing
- Audio being unplayable
- .ogg files not showing length or time elapsed. 

Improvements
- Accidental taps outside of seek bar when seeking no longer opens the file in the browser as it is blocked.
- Added loading bar so the user cannot press play before the audio is initalised.

Halkiion committed:
Fixed several bugs of the AudioPlayer plugin

Halkiion committed:
Updated version number to 1.0.7

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -34,241 +34,363 @@ import java.util.*
+            card.addView(loadingBar)
+
+            Utils.threadPool.execute {
+                val isOgg = messageAttachment.filename.lowercase(Locale.ROOT).endsWith(".ogg")
+                val localOggFile = if (isOgg) File(ctx.cacheDir, "audio.ogg") else null
+                if (isOgg) {
+                    localOggFile!!.deleteOnExit()
+                    Http.simpleDownload(messageAttachment.url, localOggFile)
+                }
```
Files should not be downloaded automatically without user interactions, and especially not like this.
IIRC `configureFileData` runs every single time the view is bound, meaning it would be re-downloaded every single time it scrolls into view. Not sure on this though.

This also creates a race condition since multiple attachments would be downloaded at the same time to the same file.

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -34,241 +34,363 @@ import java.util.*
+                }
+                val metadataPath = if (isOgg) localOggFile!!.absolutePath else messageAttachment.url
+
+                var duration: Long = try {
+                    MediaMetadataRetriever().use { retriever ->
+                        retriever.setDataSource(metadataPath)
+                        val durationStr = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
+                        durationStr?.toLong() ?: 0L
+                    }
+                } catch (e: Throwable) {
+                    0L
+                }
```
My original implementation didn't have this, but you should implement a global cache for this by url.

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -34,241 +34,363 @@ import java.util.*
+                            Utils.threadPool.execute {
+                                var playUrl = url
+                                if (isOggFile) {
+                                    val file = File(ctx.cacheDir, "audio.ogg")
+                                    file.deleteOnExit()
+                                    Http.simpleDownload(url, file)
+                                    playUrl = file.absolutePath
```
Same thing here again

rushiiMachine commented on AudioPlayer/build.gradle.kts:
```diff
@@ -1,9 +1,12 @@
-version = "1.0.6"
+version = "1.0.7"
```
```suggestion
version = "2.0.0"
```

rushiiMachine commented on AudioPlayer/build.gradle.kts:
```diff
@@ -1,9 +1,12 @@
 	changelog.set(
 		"""
+			# 1.0.7
+			* Several bug fixes fixing crashes and wrong audio clip playing
```
```suggestion
			* Fix bugs regarding crashes and desync
```

Halkiion committed:
Update AudioPlayer/build.gradle.kts

Co-authored-by: rushii <33725716+rushiiMachine@users.noreply.github.com>

Halkiion commented:
Upon doing some research, it appears that downloading the ogg file is necessary for retrieving metadata such as length using MediaMetadataRetriever. I believe having the length for ogg files is important as several users would like to use this plugin for voice messages, which are uploaded as ogg files. Would it be ok if I continue downloading the file before the user presses play so long as I implement this feature and some more cleanup logic for deleting the file once it is no longer necessary?

Alternatively, I have found a library that may offer a solution, but it may be a little too heavy in my opinion: https://github.com/wseemann/FFmpegMediaMetadataRetriever

Halkiion committed:
Add OggMetadataFetcher for fileless ogg metadata, fix more bugs, and address PR feedback

Halkiion committed:
Update AudioPlayer/build.gradle.kts

Co-authored-by: rushii <33725716+rushiiMachine@users.noreply.github.com>

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+    private fun stopCurrentPlayer() {
+
+        playerBarUpdaters.values.forEach { it() }
+        try { globalCleanup?.invoke() } catch (_: Exception) {}
+        try { globalCurrentPlayer?.setOnCompletionListener(null) } catch (_: Exception) {}
+        try { globalCurrentPlayer?.setOnPreparedListener(null) } catch (_: Exception) {}
+        try { globalCurrentPlayer?.stop() } catch (_: Exception) {}
+        try { globalCurrentPlayer?.release() } catch (_: Exception) {}
```
Is there a reason why all of these are in try/catch blocks?

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+
+        patcher.after<StoreMessages>("handleChannelSelected", Long::class.javaPrimitiveType!!) {
+            stopCurrentPlayer()
+            deleteOggCacheFiles(cacheDir = context.cacheDir, oggCacheMap = oggFileCache)
```
Don't do this when changing channels, only on start/stop (stop isn't called on app shutdown)

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+        }
+        patcher.after<AppActivity>("onCreate", Bundle::class.java) {
+            stopCurrentPlayer()
+            deleteOggCacheFiles(cacheDir = context.cacheDir, oggCacheMap = oggFileCache)
```
Same thing here, no need to do this.

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/OggMetadataFetcher.kt:
```diff
@@ -0,0 +1,110 @@
+    fun fetch(url: String): OggMetadata? {
+        var sampleRate = -1
+        var codec = ""
+        val head = fetchBytes(url, 0, 32767) ?: return null
```
Is 32KiB really necessary? I'm not familiar with these codecs but make this as little as necessary.

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/OggMetadataFetcher.kt:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    fun fetchGranulePosition(fileUrl: String): Long {
+        val tailBytes = 65536
```
Same thing here.

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
-		patcher.unpatchAll()
-	}
-}
+    fun deleteOggCacheFiles(cacheDir: File, oggCacheMap: Map<String, File>) {
+        val oggFiles = cacheDir.listFiles { file ->
+            file.parentFile == cacheDir &&
+            file.name.matches(Regex("^audio_-?\\d+\\.ogg$"))
+        } ?: emptyArray()
+        val filesToDelete = oggFiles.filter { file -> !oggCacheMap.values.contains(file) }
+        for (file in filesToDelete) {
+            file.delete()
+        }
+    }
```
You should store the cached ogg files in a subdir of the cache dir. This makes deleting it very easy with a `File#
deleteRecursively()` on `start()`. Read my other review comments about wiping cache.

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+            val ctx = root.context
+
+            card.findViewById<MaterialCardView>(playerBarId)?.let { card.removeView(it) }
+            val loadingBarId = playerBarId + 1
```
I'm not sure what you're doing here but don't do this.

rushiiMachine replied:
This is still unresolved.
Why are you increasing the ID by one? `View.generateViewId()` is a sequentially increasing ID and can create conflicts.
You should generate a separate ID.

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+            val card = root.findViewById<MaterialCardView>(attachmentCardId)
+            val ctx = root.context
+
+            card.findViewById<MaterialCardView>(playerBarId)?.let { card.removeView(it) }
+            val loadingBarId = playerBarId + 1
+            card.findViewById<ProgressBar>(loadingBarId)?.let { card.removeView(it) }
```
Don't remove views, instead set their visibility to GONE (and back to visible once needed)

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+                }
+
+                Utils.mainThread.post {
+                    card.findViewById<ProgressBar>(loadingBarId)?.let { card.removeView(it) }
```
Same thing about deleting views

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+                    card.findViewById<ProgressBar>(loadingBarId)?.let { card.removeView(it) }
+
+                    if (duration == -1L) {
+                        Toast.makeText(ctx, "Failed to load audio metadata.", Toast.LENGTH_SHORT).show()
+                        return@post
```
Don't return, allow the player to play just keep the duration at 0.
And don't show a toast, it's pointless

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+                        orientation = LinearLayout.HORIZONTAL
+                        setPadding(24, 24, 24, 24)
+                        layoutParams = FrameLayout.LayoutParams(MATCH_PARENT, WRAP_CONTENT)
+                        setOnClickListener { }
```
?

Halkiion committed:
Addressed more PR feedback

Halkiion commented:
> Could you resolve the review discussions if you've addressed them?

done

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,408 @@ import com.google.android.material.card.MaterialCardView
+    }
+
+    private fun getOggCacheDir(cacheDir: File): File {
+        val oggCacheDir = File(cacheDir, "ogg")
```
I would prefer for the cache dir to be named something more like "audio" although it doesn't really matter that much

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,408 @@ import com.google.android.material.card.MaterialCardView
+        val oggCacheDir = getOggCacheDir(cacheDir)
+        if (oggCacheDir.exists()) {
+            oggCacheDir.deleteRecursively()
+        }
```
IIRC you don't even need to check whether it exists, just ignore the result of deleteRecursively like you already do (it won't throw iirc)

rushiiMachine commented on AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt:
```diff
@@ -29,246 +29,408 @@ import com.google.android.material.card.MaterialCardView
+                                        } catch (e: Exception) {
+                                            Utils.mainThread.post {
+                                                buttonView?.isEnabled = true
+                                                Toast.makeText(ctx, "Failed to download Ogg file: ${e.localizedMessage}", Toast.LENGTH_SHORT).show()
```
```suggestion
                                                Toast.makeText(ctx, "Failed to download audio: ${e.localizedMessage}", Toast.LENGTH_SHORT).show()
```

Halkiion committed:
Update AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt

Co-authored-by: rushii <33725716+rushiiMachine@users.noreply.github.com>

Halkiion committed:
Address more PR feedback

Address more PR feedback

Address more PR feedback

Address more PR feedback

Halkiion committed:
Add request audio support for Android <8 (API level <26)

Add request audio support for Android <8 (API level <26)

Add request audio support for Android <8 (API level <26)

Add request audio support for Android <8 (API level <26)

Halkiion committed:
Add support for .opus duration

Halkiion committed:
Fix crash when switching guild/chat and optimise player code
